src/main/java/com/avispl/symphony/dal/infrastructure/epos/manager/EposManagerCommunicator.java

```diff
@@ -0,0 +1,881 @@
+		}
+		if (executorService != null) {
+			this.executorService.shutdownNow();
+			this.executorService = null;
+		}
+		if (localExtendedStatistics != null && localExtendedStatistics.getStatistics() != null) {
+			localExtendedStatistics.getStatistics().clear();
+		}
+		this.nextDevicesCollectionIterationTimestamp = 0;
+		this.aggregatedDeviceList.clear();
+		this.tenantPage = new TenantPage(100, 0, 0);
+		this.devicePage = new DevicePage(100, 0, 0);
+		this.numberOfTenant = 0;
+		this.numberOfDevice = 0;
+		this.defaultHostName = EposManagerConstant.NONE;
+		super.internalDestroy();
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	@Override
+	protected HttpHeaders putExtraRequestHeaders(HttpMethod httpMethod, String uri, HttpHeaders headers) throws Exception {
+		headers.setContentType(MediaType.APPLICATION_JSON);
+		if (loginInfo != null) {
+			headers.setBearerAuth(loginInfo.getToken());
+		}
+		return super.putExtraRequestHeaders(httpMethod, uri, headers);
+	}
+
+	/**
+	 * Verify access token is valid or expire, if token expired it will request the new one.
+	 */
+	private void checkAuthentication() throws Exception {
+		if (StringUtils.isNullOrEmpty(this.getLogin()) || StringUtils.isNullOrEmpty(this.getPassword())) {
+			throw new FailedLoginException("Username or Password field is empty. Please check device credentials");
+		}
+		if (loginInfo == null || loginInfo.isTokenExpired()) {
+			loginInfo = getLoginInfo();
+		}
+	}
+
+	/**
+	 * Retrieve authentication information by sending POST request to authentication API.
+	 */
+	private LoginInfo getLoginInfo() throws Exception {
+		LoginInfo loginInfo = null;
+		try {
+			// Set up headers
+			HttpHeaders headers = new HttpHeaders();
+			headers.setBasicAuth(this.getLogin(), this.getPassword());
+			headers.setContentType(MediaType.MULTIPART_FORM_DATA);
+
+			// Set up request body
+			MultiValueMap<String, String> body = new LinkedMultiValueMap<>();
+			body.add("grant_type", "client_credentials");
+
+			// Send POST request to device
+			HttpEntity<MultiValueMap<String, String>> request = new HttpEntity<>(body, headers);
+			String loginSubDomain = getCurrentEnvironment().getTokenGenerationSubDomain();
+
+			ResponseEntity<JsonNode> responseEntity = this.obtainRestTemplate()
+					.exchange(String.format(EposManagerUri.REQUEST_TOKEN, createRequestUrl(loginSubDomain, this.defaultHostName)), HttpMethod.POST, request, JsonNode.class);
+			JsonNode response = responseEntity.getBody();
+
+			if (response == null || !response.has(EposManagerConstant.ACCESS_TOKEN)) {
+				throw new ResourceNotReachableException("Failed to retrieve authentication information, endpoint not reachable");
+			}
+
+			String token = response.get(EposManagerConstant.ACCESS_TOKEN).asText();
+			long expiresIn = response.get(EposManagerConstant.EXPIRES_IN).asLong();
+			loginInfo = new LoginInfo(token, expiresIn);
+
+		} catch (HttpClientErrorException e) {
+			if (e.getStatusCode() == HttpStatus.BAD_REQUEST) {
```

Comment:
please add alternative case here. What if it isn't badrequest specifically - the error will get swallowed